cpp/include/raft/neighbors/detail/nn_descent_batch.cuh

```diff
@@ -0,0 +1,684 @@
+/*
+ * Copyright (c) 2023-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <raft/util/cudart_utils.hpp>
+#undef RAFT_EXPLICIT_INSTANTIATE_ONLY
+
+#include "../nn_descent_types.hpp"
+#include "nn_descent.cuh"
+
+#include <raft/cluster/kmeans_balanced.cuh>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/managed_mdarray.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/mdspan_types.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/matrix/detail/gather_inplace.cuh>
+#include <raft/matrix/init.cuh>
+#include <raft/matrix/sample_rows.cuh>
+#include <raft/neighbors/brute_force-inl.cuh>
+
+#include <cub/cub.cuh>
+
+#include <vector_types.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <limits>
+#include <optional>
+#include <random>
+#include <type_traits>
+
+#define INST_MERGE_SUBGRAPH(blockSize, elemPT)                                                 \
```

Comment:
I can't figure out why this was needed? The kernel is already a templated function?

Reply:
Oh it's because it's so long I didn't want to write it multiple times in the if-else below where we launch the kernel..!
referred to what is being done [here](https://github.com/rapidsai/raft/blob/32f3703b0d28e5e8934fc463cf3ec6bd6cc82a79/cpp/include/raft/matrix/detail/columnWiseSort.cuh#L216)

Reply:
Let's not use macros, please. It's fine to have the kernel invocation verbosity as it makes the code more readable.

Reply:
Okay 🙂